src/main/java/org/apache/flink/benchmark/SortingBoundedInputBenchmarks.java

```diff
@@ -0,0 +1,418 @@
+		DataStreamSource<Tuple2<Integer, byte[]>> elements2 = env.fromParallelCollection(
+			new InputGenerator(RECORDS_PER_INVOCATION / 2),
+			new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)
+		);
+		SingleOutputStreamOperator<Long> counts = elements1.connect(elements2)
+			.keyBy(
+				element -> element.f0,
+				element -> element.f0
+			)
+			.transform(
+				"Asserting operator",
+				BasicTypeInfo.LONG_TYPE_INFO,
+				new AssertingTwoInputOperator()
+			);
+
+		counts.addSink(new DiscardingSink<>());
+	}
+
+	@Benchmark
+	@OperationsPerInvocation(value = RECORDS_PER_INVOCATION)
+	public void testThreeInputOperator(SortingInputContext context) {
+		StreamExecutionEnvironment env = context.env;
+
+		KeyedStream<Tuple2<Integer, byte[]>, Object> elements1 = env.fromParallelCollection(
+			new InputGenerator(RECORDS_PER_INVOCATION / 3),
+			new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)
+		).keyBy(el -> el.f0);
+
+		KeyedStream<Tuple2<Integer, byte[]>, Object> elements2 = env.fromParallelCollection(
+			new InputGenerator(RECORDS_PER_INVOCATION / 3),
+			new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)
+		).keyBy(el -> el.f0);
+
+		KeyedStream<Tuple2<Integer, byte[]>, Object> elements3 = env.fromParallelCollection(
+			new InputGenerator(RECORDS_PER_INVOCATION / 3),
+			new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)
+		).keyBy(el -> el.f0);
+
+		KeyedMultipleInputTransformation<Long> assertingTransformation = new KeyedMultipleInputTransformation<>(
+			"Asserting operator",
+			new AssertingThreeInputOperatorFactory(),
+			BasicTypeInfo.LONG_TYPE_INFO,
+			-1,
+			BasicTypeInfo.INT_TYPE_INFO
+		);
+		assertingTransformation.addInput(elements1.getTransformation(), elements1.getKeySelector());
+		assertingTransformation.addInput(elements2.getTransformation(), elements2.getKeySelector());
+		assertingTransformation.addInput(elements3.getTransformation(), elements3.getKeySelector());
+
+		env.addOperator(assertingTransformation);
+		DataStream<Long> counts = new DataStream<>(env, assertingTransformation);
+
+		counts.addSink(new DiscardingSink<>());
+	}
+
+	private static class AssertingOperator extends AbstractStreamOperator<Long>
+			implements OneInputStreamOperator<Tuple2<Integer, byte[]>, Long>, BoundedOneInput {
+		private final Set<Integer> seenKeys = new HashSet<>();
+		private long seenRecords = 0;
+		private Integer currentKey = null;
+
+		@Override
+		public void processElement(StreamRecord<Tuple2<Integer, byte[]>> element) throws Exception {
+			this.seenRecords++;
+			Integer incomingKey = element.getValue().f0;
+			if (!Objects.equals(incomingKey, currentKey)) {
+				if (!seenKeys.add(incomingKey)) {
+					Assert.fail("Received an out of order key: " + incomingKey);
+				}
+				this.currentKey = incomingKey;
+			}
+		}
```

Comment:
nit: maybe extract this code to a helper class `ProcessedKeysOrderAsserter` and re-use it for 1, 2 and 3 input operators?